src/python/pants/backend/python/tasks/pex_build_util.py

```diff
@@ -46,6 +46,59 @@ def has_python_requirements(tgt):
   return isinstance(tgt, PythonRequirementLibrary)
 
 
+def is_python_binary(tgt):
+  return isinstance(tgt, PythonBinary)
+
+
+def tgt_closure_has_native_sources(tgts):
+  """Determine if any target in the current target closure has native (c or cpp) sources."""
+  return any(tgt.has_native_sources for tgt in tgts)
+
+
+def tgt_closure_platforms(tgts):
+  """
+  Aggregates a dict that maps a platform string to a list of targets that specify the platform.
+  If no targets have platforms arguments, return a dict containing platforms inherited from
+  the PythonSetup object.
+
+  :param tgts: a list of :class:`Target` objects.
+  :returns: a dict mapping a platform string to a list of targets that specify the platform.
+  """
+  tgts_by_platforms = {}
+  for tgt in tgts:
+    if tgt.platforms:
+      for platform in tgt.platforms:
+        if platform in tgts_by_platforms:
+          tgts_by_platforms[platform].append(tgt)
+        else:
+          tgts_by_platforms[platform] = [tgt]
+  # If no targets specify platforms, inherit the default platforms.
+  if not tgts_by_platforms:
+    for platform in PythonSetup.global_instance().platforms:
+      tgts_by_platforms[platform] = ['(No target) Platform inherited from either the '
+                                     '--platforms option or a pants.ini file.']
+  return tgts_by_platforms
+
+
+def build_for_current_platform_only_check(tgts):
+  """
+  Performs a check of whether the current target closure has native sources and if so, ensures that
+  Pants is only targeting the current platform.
+
+  :param tgts: a list of :class:`Target` objects.
+  :return: a boolean value indicating whether the current target closure has native sources.
+  """
+  if tgt_closure_has_native_sources(filter(is_local_python_dist, tgts)):
+    platforms = tgt_closure_platforms(filter(is_python_binary, tgts))
+    if len(platforms.keys()) > 1 or not 'current' in platforms.keys():
+      raise IncompatiblePlatformsError('The target set contains one or more targets that depend on '
+        'native code. Please ensure that the platform arguments in all relevant targets and build '
+        'options are compatible with the current platform. Found targets for platforms: {}'
```

Comment:
The idea here was that we could actually show one or more of the incompatible targets, but let's just save that for a future change.

Reply:
`platforms` isn't a great name for this var. Since it is a dict, the output will contain a string repr of keys-> list of targets w/ addresses when this exception is raised.

Reply:
Ooh, right. :thumbsup: